### PR TITLE
Cleanups and modernization

### PR DESCRIPTION
**Pull request by nbougalis**
Reviews: @rec, @scottschurr

### CODE REVIEW

**Comment by scottschurr** [comment]
@nbougalis, you saw the compile error from Travis using Clang?  Looks legit.

**Comment by scottschurr** [comment]
there are a lot of really good cleanups in this work.  Thanks for doing it.

Before giving the thumbs up on this commit, I'd like to be convinced of the correctness of three things (in priority order):
 1. That the explicit conversion operator in STVector256 is better than if it were a named method.
 2. That `duplicate` is a better name than `clone` in the `STBase` inheritance hierarchy, and
 3. That the `duplicate` / `clone` method of `STBase` should not be pure virtual.

Thanks.

**Comment by nbougalis** [comment]
First, a little sidebar. The code previously had two functions:
    
    std::vector<uint256> const& peekValue () const;
    std::vector<uint256>& peekValue();

The changes eliminated the need for the non-const version, which left us only with the version returning the constant reference. That version was called only in one place where we took the result and copy-constructed a new `std::vector<uint256>`. So I then replaced that function with an explicit `operator std::vector<uint256>` which returns a copy of of the contained `std::vector` to begin with.

Now, on to the issue of using `static_cast` which @scottschurr mentioned:

As @HowardHinnant pointed out, the uniform syntax is a big advantage. That it's not recommended in a book or accepted as a mainstream practice isn't really a concern to me.

I personally prefer the `static_cast` because it lets me express the intent of the code and, once written, it's self-documenting and its semantics are crystal clear. It boils down to this: "I have an X and I need a Y. Do it." I don't have to go check the available functions and decide which one is best suited - I can just ask for what I want and, if it's possible, then I just get it.

Do note that the `explicit` conversion operator will never be called unless a programmer requests a conversion, much like with a function, so you can't get into weird situations where things get converted unknowningly.

**Comment by scottschurr** [comment]
Okay.  I'm sold on the explicit conversion operator.  I had expectations that `static_cast` ought to be a light weight operation.  I've now been disabused of that illusion.  I'll need to train myself to think yet harder when I see `static_cast`.

My personal preference will continue to lean toward named methods (unless I know I need to be generic).  But I won't inflict that personal preference on code that others write.

**Comment by nbougalis** [comment]
@scottschurr I'm not sure that my personal preferences should have more weight than yours for this codebase.

Does anyone else want to chime in? @vinniefalco, @rec, @ximinez, @josh-ripple?

**Comment by vinniefalco** [comment]
Its not a problem until it becomes a problem. And it hasn't become a problem. Yet. Let's see how it goes.

**Comment by HowardHinnant** [comment]
I must admit I got nervous when I realized that Nik was replacing return by reference with return by value.  But I should've known better than to worry.  In all the places I spot checked, Nik did not insert a performance penalty.  Nice job Nik! :+1:

**Comment by josh-ripple** [comment]
I think of static_cast as indicating a cast from void* or upcasts in general.  I personally wouldn't use it when construction syntax is an option.

**Comment by josh-ripple** [comment]
Okay, construction syntax is not an option here.  I would write 
```    std::vector<uint256> const& get() const
    {
        return mValue;
    }
```
in class STVector256 and `sleAmendments->getFieldV256 (sfAmendments).get()` in Ledger::getLedgerAmendments().

Also, I suggest using two return statements instead of a temporary std::vector, to take advantage of Return Value Optimization (RVO).

**Comment by nbougalis** [comment]
So, general consensus on this?

**Comment by scottschurr** [comment]
:+1:
I'm not excited about the explicit conversion operator or replacing `clone()` with `duplicate()`.  But those are minor readability issues (to me).  They should work just fine.

I'm surprised that I haven't seen any comments from @rec yet...

File: src/ripple/app/ledger/AcceptedLedgerTx.cpp [hunk 1]
```diff
@@ -84,7 +84,7 @@ void AcceptedLedgerTx::buildJson ()
     if (!mAffected.empty ())
     {
         Json::Value& affected = (mJson[jss::affected] = Json::arrayValue);
-        BOOST_FOREACH (const RippleAddress & ra, mAffected)
+        for(auto const& ra : mAffected)
```
Comment (rec) [inline src/ripple/app/ledger/AcceptedLedgerTx.cpp]: Space between `for` and `(auto` please!

**Comment by ximinez** [comment]
I don't really have an opinion on he explicit cast vs. the `peekValue` function. Try it and see what happens. If it causes trouble, we can readdress.

**Comment by nbougalis** [comment]
Apologies to reviewers, but I had to rebase.

**Comment by rec** [comment]
OK, I think this is pretty well ready to go.

Can you rebase this and I'll make a final review _just for correctness and not for style or optimality?_

**Comment by nbougalis** [comment]
@rec This is already rebased. Is there something specific you'd like, or would you prefer I squash it down to final form?

@scottschurr I didn't do the `duplicate` to `clone` rename, but my work on the serialized types parts isn't over. I like the idea and will definitely implement it in the future.

File: src/ripple/app/peers/UniqueNodeList.cpp [hunk 2]
```diff
@@ -813,8 +823,8 @@ class UniqueNodeListImp
     //
     void scoreCompute ()
     {
-        strIndex                umPulicIdx;     // Map of public key to index.
-        strIndex                umDomainIdx;    // Map of domain to index.
+        hash_map<std::string, int> umPulicIdx;     // Map of public key to index.
```
Comment (scottschurr) [inline src/ripple/app/peers/UniqueNodeList.cpp]: Ultra-minor nit (and not your change): I think `umPulicIdx` is supposed to be `umPublicIdx`.

File: src/ripple/protocol/STBase.h [hunk 3]
```diff
@@ -126,9 +127,6 @@ class STBase
     SField::ref
     getFName() const;
 
-    std::unique_ptr<STBase>
-    clone() const;
```
Comment (scottschurr) [inline src/ripple/protocol/STBase.h]: The reason I was nattering about `clone` is that all of the base classes used to get this (non-virtual) `clone` interface and implementation.  Internally that `clone` implementation called the private virtual `duplicate` methods.  You got rid of the indirection from `clone` to `duplicate`.  That's great.  Anyway, that's why I've been saying that the name changed from `clone` to `duplicate`.  You're correct, the old `duplicate` methods were there before.  But the public access to those methods was through a `clone` method that's now gone.

I'm fine with your decision to postpone the rename.  I just thought I'd try to explain myself a bit more clearly.

**Comment by scottschurr** [comment]
:+1:
Saw no errors introduced by the rebase.  It's good to go as far as I can see.

**Comment by rec** [comment]
I just wanted to see the final rebase!  Each change is good for me.

:+1:

**Comment by nbougalis** [comment]
Merged as 47593730d684ba88f238bf67d9d0bd0f11166f76

**Pull request closed without merging**